README.md
# RS

Envía comprobantes de las facturas insertadas en la base de datos a la API de facturas de REDABITS.

# Instalación

Primero lea esta [guía de instalación de ODBC para Windows](https://docs.microsoft.com/en-us/sql/connect/python/pyodbc/step-1-configure-development-environment-for-pyodbc-python-development?view=sql-server-2017) hasta el punto 3. Luego de descargar el repositorio se ejecuta `pip` para instalar el paquete con todas sus dependencias:

```
pip install .
```

# Variables de entorno

- `DEBUG`: Si esta establecido (con cualquier valor diferente a una cadena vacía) usa una configuración adecuada para pruebas. **No se meten datos falsos a ninguna base de datos durante la ejecución**. Para usar valores de producción esta variable de entorno **no** debe ser establecida, además, debe correr el script con la bandera `--production`.
- `DATABASE_URL`: Indica la URL o el DSN de la base de datos MSSQL. El formato para DSN tiene que ser `mssql+pyodbc://<username>:<password>@<dsnname>`, para conexion por URL el formato debe ser `mssql+pyodbc://<username>:<password>@<host>:<port>/<databasename>?driver=<SQL+Server+Native+Client+XX.YY>` el SQL Server driver puede variar, para conocer el valor correcto se puede usar el siguiente comando desde la linea de comandos:

```
python -c 'import pyodbc; print(pyodbc.drivers())'
```

# Como usar

Simplemente ejecute `rsb --production` para una ejecución con configuración de producción.

# Pruebas de integración (CAMBIOS DESTRUCTIVOS EN LA BASE DE DATOS)

**Las pruebas unitarias deben realizarse en una base de datos que no sea del cliente**. Dentro del proyecto puede ejecutar `python setup.py test`.


CHANGELOG.rst
Changelog
=========


0.3 (2018-09-19)
----------------

Changes
~~~~~~~
- Show basic information for class instance of Factura. [Jorge Araya
  Navarro]

Fix
~~~
- Update Factura counter during testing. [Jorge Araya Navarro]

  Avoid sending repeated voucher number to Hacienda


0.2 (2018-09-19)
----------------

Changes
~~~~~~~
- Bump version. [Jorge Araya Navarro]
- Add Impuesto only if there is some. [Jorge Araya Navarro]
- Update RESUMEN_* columns in Comprobante model. [Jorge Araya Navarro]

  In accordance with the following JSON example:

      "ResumenFactura":
  	{
  		"CodigoMoneda":"",//USD para dolares, CRC para colones. Si es null por defecto es CRC
  		"TipoCambio":"",//Obligatorio si "CodigoMoneda" != CRC
  		"TotalMercanciasGravadas":2654.88,//suma de "SubTotal" de las lineas que tengan "Impuesto"
  		"TotalMercanciasExentas":0,//suma de "SubTotal" de las lineas que no tengan "Impuesto"
  		"TotalGravado":2654.88,//igual que "TotalMercanciasGravadas" en este proyecto
  		"TotalExento":0,//igual que "TotalMercanciasExentas" en este proyecto
  		"TotalVenta":2654.88,//"TotalGravado" + "TotalExento"
  		"TotalDescuentos":0,//suma de "MontoDescuento" de las lineas que lo tengan
  		"TotalVentaNeta":2654.88,//"TotalVenta" - "TotalDescuentos"
  		"TotalImpuesto":345.1344,//suma de "Monto" de todos los "Impuesto"
  		"TotalComprobante":3000.0144//"TotalVentaNeta" + "TotalImpuesto"
  	}
- Raise error in status 200 with bogus content. [Jorge Araya Navarro]

  Prevents quirks from the API by ensuring that clave is a long string of digits.

Fix
~~~
- Fix wrong access to dict inside list. [Jorge Araya Navarro]
- Include missing field. [Jorge Araya Navarro]
- Use IV column for taxes instead of hardcoded amount. [Jorge Araya
  Navarro]
- Check if clave is form out of digits. [Jorge Araya Navarro]

  Avoid total failure of the app
- Use same URL base for testing. [Jorge Araya Navarro]
- Use different URL base for sandbox. [Jorge Araya Navarro]
- Remove unused imports. [Jorge Araya Navarro]
- Add field as a list of dicts. [Jorge Araya Navarro]
- Add missing field and correction of typos. [Jorge Araya Navarro]
- Fix misnamed field in payload. [Jorge Araya Navarro]

Other
~~~~~
- Fixing the conversion of the num_factura. [Dennis Hernández]


0.1 (2018-09-18)
----------------

New
~~~
- Change in README files. [Jorge Araya Navarro]
- Update Comprobante state when receipt is received by API. [Jorge Araya
  Navarro]
- Add executable script. [Jorge Araya Navarro]

  This script does the job of sending all receipts in batches.
- Add model for table, functions and new tests. [Jorge Araya Navarro]

  - https://ptpb.pw/GaBt
- Implementa modelo para varias tablas. [Jorge Araya Navarro]

  Las tablas afectadas son `FACTURAS`, `HISTORICO_MOV`, `COMPROBANTE` y `COMPROBANTE_LINEA_DETALLE`
- Provide object for connection with MSSQL database. [Jorge Araya
  Navarro]

  `DATABASE_URL` is an environment variable which holds the information required to connect with a
  database server.
- Add Python dependencies. [Jorge Araya Navarro]
- Change docker service allowing database restore. [Jorge Araya Navarro]

  Binding directories allow for the Docker service to reach the backup file
- Enable usage of Docker for development. [Jorge Araya Navarro]

  MSSQL image for docker can be use for a zero-conf MSSQL database
- Add README. [Jorge Araya Navarro]
- Add setup.py. [Jorge Araya Navarro]
- Aplica directiva de archivos ignorados para Git. [Jorge Araya Navarro]

Changes
~~~~~~~
- Format source code and remove models. [Jorge Araya Navarro]
- Format source code with Black. [Jorge Araya Navarro]

  - https://github.com/ambv/black
- Borra prueba unitaria. [Jorge Araya Navarro]
- Factura model generate the purchase details. [Jorge Araya Navarro]
- Remove ComprobanteDetalle model. [Jorge Araya Navarro]
- Several changes for models. [Jorge Araya Navarro]
- Set Clave as returned by Daniel's API. [Jorge Araya Navarro]
- Add Emisor data from API request. [Jorge Araya Navarro]

  - commit b3f256f70bb83558ab5c0f119b8d3ae7429cdcd1
  - issue #5
- Update integration test. [Jorge Araya Navarro]
- Ensure field has content. [Jorge Araya Navarro]
- Generate several invoices for testing. [Jorge Araya Navarro]

  Boilerplate data from the database that help with integration tests
- Add debugging information for API consumer. [Jorge Araya Navarro]

  Sometimes the API consumed does not return expected responses
- Establece relacion entre tablas. [Jorge Araya Navarro]

  Tablas afectadas son `COMPROBANTE` y `COMPROBANTE_LINEA_DETALLE`, la relación es *one-to-many*
  siendo `COMPROBANTE` el padre.

  - https://docs.sqlalchemy.org/en/latest/orm/basic_relationships.html#one-to-many
- Explain where scripts should go. [Jorge Araya Navarro]

Fix
~~~
- Fix length of field in table Comprobante. [Jorge Araya Navarro]

  - fixes issue #6
- Fix column length and add missing fields. [Jorge Araya Navarro]
- Return requests response. [Jorge Araya Navarro]
- Ensure error are raised if `DetallesServicio` is empty. [Jorge Araya
  Navarro]
- Remove several fields from model. [Jorge Araya Navarro]

  They do not exists in the new SQL scripts

  - fix issue #4
- Fix boilerplate data generator. [Jorge Araya Navarro]
- Fix wrong syntax. [Jorge Araya Navarro]

  - fixes issue #2
- Arregla argumento posicional luego de argumento nombrado. [Jorge Araya
  Navarro]

  ```
      Integer, name='COMPROBANTE_ID', ForeignKey('COMPROBANTE.ID'))
                                     ^
  SyntaxError: positional argument follows keyword argument
  ```
- Use correct dependency versions. [Jorge Araya Navarro]
- Move the directory bind outside project directory. [Jorge Araya
  Navarro]

  There is interference with `pip install` because the file permission of the database data under `data/db/`
- Include missing directory. [Jorge Araya Navarro]

Other
~~~~~
- Removing NOT NULL. [Dennis Hernández]
- Deleting linea_detalle and adding NUM_FACTURA to Comprobante.
  [djhvscf]

rs/test/testmapeo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
""" Prueba si el mapeo de modelos a tablas funciona en MSSQL.
"""

import unittest

from rs import session
from rs.models import Factura, Historico


class TestMapeo(unittest.TestCase):
    def test_factura(self):
        """ Consigue cualquier cosa de la tabla FACTURAS
        """
        obj = session.query(Factura).filter(Factura.num_factura == 35.0).all()

    def test_historico(self):
        """ Consigue cualquier cosa de la tabla HISTORICO_MOV
        """
        obj = session.query(Historico).filter(
            Historico.num_factura == 602.0).all()


rs/test/testintegration.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import os
import unittest

from rs import session
from rs.models import Factura


class TestIntegration(unittest.TestCase):
    def setUp(self):
        # Use sandbox/testing capabilities on all other third-party services
        os.environ["DEBUG"] = "ON"

        # Create some invoices, this should trigger the SP and make new
        # receipts. WARNING: There is no rollback between tests, so changes in
        # the tables are accumulative, thus, it may take disk space or add
        # trash to production database if used for testing.
        Factura()._populate()

    def tearDown(self):
        session.rollback()

    def test_send(self):
        """ Send a recipe and see what happens
        """
        from rs.funcs import sendall

        sendall()


rs/funcs.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import logging
import tqdm
from rs import session
from rs.consumer import send
from rs.models import Comprobante

CLAVE_URL = "http://api.redabits.com/getrespuesta.php?clave=%s"


def receipts():
    """ Return a list of pending receipts

    They are batched in batch of 100 rows
    """
    for receipt in (
        session.query(Comprobante)
        .yield_per(100)
        .filter(Comprobante.enviado == False)
        .all()
    ):
        yield receipt


def sendall():
    """ Send all pending receipts
    """
    logging.info("Sending pending receipts...")
    total = Comprobante.count()
    for comprobante in tqdm.tqdm(receipts(), total=total, ascii=True):
        resp = send(comprobante)
        clave = resp.content.decode("utf-8").strip()
        if clave.isdigit():
            comprobante.enviado = True
            comprobante.clave = clave
            factura = comprobante.get_factura()
            factura.folio = clave
            factura.enlace = CLAVE_URL + clave
            session.commit()
        else:
            # Problemas con el servidor
            logging.error(
                "Could not get accepted receipt #%s. API response: %s",
                comprobante.numero_consecutivo,
                clave,
            )
            logging.info(
                "Will try re-sending receipt #%s on next run",
                comprobante.numero_consecutivo,
            )


rs/consumer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import logging
from os import getenv
from urllib.parse import urljoin

import requests
from rs.models import Comprobante

AUTH_API_URL = "http://api.redabits.com/credentials.php"
RECIPE_API_URL = "http://api.redabits.com:3000/factura"
MH_API_BASE = "https://api.comprobanteselectronicos.go.cr/"
MH_API_TEST_BASE = "https://idp.comprobanteselectronicos.go.cr/"
MH_API_PROD = urljoin(MH_API_BASE, "recepcion/v1")
MH_API_TEST = urljoin(MH_API_BASE, "recepcion-sandbox/v1")
MH_API_AUTH_TEST = urljoin(
    MH_API_TEST_BASE, ("auth/realms/rut-stag/protocol/openid-connect/token")
)
MH_API_AUTH_PROD = urljoin(
    MH_API_BASE, ("auth/realms/rut/protocol/openid-connect/token")
)

if getenv("DEBUG"):
    logging.basicConfig()
    logging.getLogger().setLevel(logging.DEBUG)
    requests_log = logging.getLogger("requests.packages.urllib3")
    requests_log.setLevel(logging.DEBUG)
    requests_log.propagate = True


def auth(comprobante: Comprobante) -> dict:
    """ Ask and set authentication data into the payload
    """
    payload = comprobante.auth_data()
    req = requests.post(AUTH_API_URL, data=payload, timeout=30)
    logging.debug(req.content)
    json = req.json()
    data = comprobante.marshall()
    docendp = MH_API_TEST if getenv("DEBUG") else MH_API_PROD
    authendp = MH_API_AUTH_TEST if getenv("DEBUG") else MH_API_AUTH_PROD
    # Set authentication data
    data["auth"] = {
        "usuario": json["usuario"],
        "password": json["password"],
        "cert": json["cert"],
        "pin": json["pin"],
        "documents_endpoint": docendp,
        "authentication_endpoint": authendp,
        "api_client_id": "api-stag" if getenv("DEBUG") else "api-prod",
    }
    # Set Emisor data
    data["Emisor"]["Nombre"] = json["nombre"]
    data["Emisor"]["Identificacion"]["Tipo"] = json["tipo_id"]
    data["Emisor"]["Telefono"]["NumTelefono"] = json["telefono"]
    data["Emisor"]["CorreoElectronico"] = json["correo"]

    return data


def send(comprobante: Comprobante):
    """ Send the recipe to redabits API server
    """
    payload = auth(comprobante)
    req = requests.post(RECIPE_API_URL, json=payload, timeout=30)
    logging.debug(req.content)
    return req


rs/models.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from os import getenv

from rs import session
from sqlalchemy import Boolean, Column, DateTime, Float, Integer, SmallInteger, String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.types import DECIMAL
from sqlalchemy.orm.exc import MultipleResultsFound

BASE = declarative_base()


class Factura(BASE):  # type: ignore
    __tablename__ = "FACTURAS"

    num_factura = Column(Float, primary_key=True, name="NUM_FACTURA")
    cod_pos = Column(String(8), primary_key=True, name="COD_POS")
    tipo = Column(String(11), primary_key=True, name="TIPO")
    fecha = Column(DateTime, nullable=True, name="FECHA")
    monto_total = Column(Float, nullable=True, name="MONTO_TOTAL")
    subtotal = Column(Float, nullable=True, name="SUBTOTAL")
    monto_contado = Column(Float, nullable=True, name="MONTO_CONTADO")
    monto_cheque = Column(Float, nullable=True, name="MONTO_CHEQUE")
    monto_tarjeta = Column(Float, nullable=True, name="MONTO_TARJETA")
    monto_credito = Column(Float, nullable=True, name="MONTO_CREDITO")
    monto_nc = Column(Float, nullable=True, name="MONTO_NC")
    monto_certificado = Column(Float, nullable=True, name="MONTO_CERTIFICADO")
    iv = Column(Float, nullable=True, name="IV")
    vendedor = Column(String(50), nullable=True, name="VENDEDOR")
    cajero = Column(String(50), nullable=True, name="CAJERO")
    cliente = Column(String(50), nullable=True, name="CLIENTE")
    descuento = Column(Float, nullable=True, name="DESCUENTO")
    cod_tarjeta = Column(String(6), nullable=True, name="COD_TARJ")
    observaciones = Column(String(100), nullable=True, name="OBSERVACIONES")
    hilera = Column(String(2000), nullable=True, name="HILERA")
    mensaje_error = Column(String, nullable=True, name="MENSAJE_ERROR")
    cedula = Column(String(20), nullable=True, name="CEDULA")
    cedula_tipo = Column(String(2), nullable=True, name="TIPO_CEDULA")
    correo = Column(String(100), nullable=True, name="CORREO")
    folio = Column(String(250), nullable=True, name="FOLIOMH")
    enlace = Column(String(500), nullable=True, name="LINK")
    enlace_respuesta = Column(String(500), nullable=True, name="LINK_RESPUESTA")
    pdf = Column(String(500), nullable=True, name="LINK_PDF")
    respuesta_hacienda = Column(String(1000), nullable=True, name="RESPUESTA_HACIENDA")
    respuesta_tributacion = Column(
        String(1000), nullable=True, name="RESPUESTA_TRIBUTACION"
    )

    def __repr__(self):
        return "<Factura {} folio {}>".format(self.num_factura, self.folio)

    def get_details(self):
        """ Return a list of purchase details
        """
        linea = 1

        for detail in (
            session.query(Historico)
            .filter(
                Historico.cod_pos == self.cod_pos,
                Historico.num_factura == self.num_factura,
            )
            .all()
        ):
            montototal = detail.cantidad * detail.precio
            montodescuento = montototal * (self.descuento / self.subtotal)
            descuentodesc = "Descuentos otorgados" if montodescuento else ""
            subtotal = montototal - montodescuento
            if detail.iv is None:
                detail.iv = 0
            data = {
                "LineaDetalle": {
                    "NumeroLinea": linea,
                    "Codigo": {"Tipo": "04", "Codigo": detail.codigo.strip()},
                    "Cantidad": round(detail.cantidad, 3),
                    "UnidadMedida": "Unid",
                    "Detalle": detail.desc_producto,
                    "PrecioUnitario": round(detail.precio, 5),
                    "MontoTotal": round(montototal, 5),
                    "MontoDescuento": round(montodescuento, 5),
                    "NaturalezaDescuento": descuentodesc,
                    "SubTotal": round(subtotal, 5),
                    "MontoTotalLinea": round(
                        subtotal + (subtotal * (detail.iv / 100)), 5
                    ),
                }
            }
            if detail.iv > 0:
                data["LineaDetalle"]["Impuesto"] = [
                    {
                        "Codigo": "01",
                        "Tarifa": detail.iv,
                        "Monto": round(subtotal * (detail.iv / 100), 5),
                    }
                ]
            linea = linea + 1
            yield data

    @staticmethod
    def _populate(amount=200):
        """ Add several invoices into the database for testing purpose.
        """
        from mimesis import Generic
        from rs import session
        from datetime import datetime

        g = Generic("es")

        r = session.execute("SELECT TOP 1 [FACTURA_ACTUAL] FROM [dbo].[POS]").first()[0]

        for n in range(int(r), int(r) + amount):
            factura = Factura(
                num_factura=n,
                cod_pos="TIENDA",
                tipo="F",
                monto_total=n,
                subtotal=n,
                monto_contado=n,
                monto_cheque=0,
                monto_tarjeta=0,
                monto_credito=0,
                monto_nc=0,
                monto_certificado=0,
                iv=0,
                vendedor=g.person.full_name(),
                cajero=g.person.full_name(),
                cliente=g.person.full_name(),
                descuento=0,
                cedula="114760094",  # de Daniel
                correo=g.person.email(),
            )
            session.add(factura)
            session.execute(
                "UPDATE [dbo].[POS] SET [FACTURA_ACTUAL] = [FACTURA_ACTUAL] + 1"
            )
            for detalle in range(5):
                historico = Historico(
                    num_factura=n,
                    cod_pos="TIENDA",
                    tipo="F",
                    codigo="100001",
                    fecha=datetime.now(),
                    cantidad=5,
                    precio=1000,
                    desc_producto=g.food.dish(),
                    proveedor="GENERAL",
                    departamento="IT",
                    costo=0,
                    num_bodega=1,
                    iv=13,
                )
                session.add(historico)
            session.commit()


class Historico(BASE):  # type: ignore
    __tablename__ = "HISTORICO_MOV"

    num_factura = Column(Float, primary_key=True, name="NUM_FACTURA")
    cod_pos = Column(String(8), primary_key=True, name="COD_POS")
    tipo = Column(String(1), primary_key=True, name="TIPO")
    codigo = Column(String(25), primary_key=True, name="CODIGO")
    fecha = Column(DateTime, primary_key=True, name="FECHA")
    cantidad = Column(Float, nullable=True, name="CANTIDAD")
    precio = Column(Float, nullable=True, name="PRECIO")
    desc_producto = Column(String(55), nullable=True, name="DESCRIPCION_PRODUCTO")
    departamento = Column(String(55), nullable=True, name="DEPARTAMENTO")
    proveedor = Column(String(55), nullable=True, name="PROVEEDOR")
    costo = Column(Float, nullable=True, name="COSTO")
    num_bodega = Column(SmallInteger, nullable=True, name="NUM_BODEGA")
    observaciones = Column(String(2000), nullable=True, name="OBSERVACIONES")
    iv = Column(Float, nullable=True, name="IV")

    def __repr__(self):
        return "<Historico {} ({})>".format(self.num_factura, self.cod_pos)


class Comprobante(BASE):  # type: ignore
    __tablename__ = "COMPROBANTE"

    id = Column(Integer, primary_key=True, name="ID")
    clave = Column(String, name="CLAVE")
    num_factura = Column(Float, name="NUM_FACTURA")
    numero_consecutivo = Column(String, name="NUMERO_CONSECUTIVO")
    fecha_emision = Column(DateTime, name="FECHA_EMISION")
    condicion_venta = Column(String(2), name="CONDICION_VENTA")
    medio_pago = Column(String(2), name="MEDIO_PAGO")
    emisor_ident_num = Column(String(50), name="EMISOR_IDENT_NUM")
    emisor_provincia = Column(String(2), name="EMISOR_PROVINCIA")
    emisor_canton = Column(String(3), name="EMISOR_CANTON")
    emisor_distrito = Column(String(3), name="EMISOR_DISTRITO")
    emisor_otras_senas = Column(String(500), name="EMISOR_OTRAS_SENAS")
    receptor_nombre = Column(String(100), nullable=True, name="RECEPTOR_NOMBRE")
    receptor_ident_tipo = Column(String(2), nullable=True, name="RECEPTOR_IDENT_TIPO")
    receptor_ident_num = Column(String(2), nullable=True, name="RECEPTOR_IDENT_NUM")
    resumen_cod_moneda = Column(String(3), name="RESUMEN_COD_MONEDA")
    resumen_tipo_cambio = Column(
        DECIMAL(precision=18, scale=2), nullable=True, name="RESUMEN_TIPO_CAMBIO"
    )
    resumen_total_mercancias_gravadas = Column(
        DECIMAL(precision=18, scale=2), name="RESUMEN_TOTAL_MERCANCIAS_GRAVADAS"
    )
    resumen_total_mercancias_exentas = Column(
        DECIMAL(precision=18, scale=2), name="RESUMEN_TOTAL_MERCANCIAS_EXENTAS"
    )
    resumen_total_gravado = Column(
        DECIMAL(precision=18, scale=2), name="RESUMEN_TOTAL_GRAVADO"
    )
    resumen_total_exento = Column(
        DECIMAL(precision=18, scale=2), name="RESUMEN_TOTAL_EXENTO"
    )
    resumen_total_venta = Column(
        DECIMAL(precision=18, scale=2), name="RESUMEN_TOTAL_VENTA"
    )
    resumen_total_descuentos = Column(
        DECIMAL(precision=18, scale=2), name="RESUMEN_TOTAL_DESCUENTOS"
    )
    resumen_total_venta_neta = Column(
        DECIMAL(precision=18, scale=2), name="RESUMEN_TOTAL_VENTA_NETA"
    )
    resumen_total_impuesto = Column(
        DECIMAL(precision=18, scale=2), name="RESUMEN_TOTAL_IMPUESTO"
    )
    resumen_total_comprobante = Column(
        DECIMAL(precision=18, scale=2), name="RESUMEN_TOTAL_COMPROBANTE"
    )
    referencia_numero = Column(String(100), name="REFERENCIA_NUMERO")
    normativa_num_resolucion = Column(String(100), name="NORMATIVA_NUM_RESOLUCION")
    normativa_fecha_resolucion = Column(String(100), name="NORMATIVA_FECHA_RESOLUCION")
    otros = Column(String(500), nullable=True, name="OTROS")
    notify_to = Column(String(500), name="NOTIFY_TO")
    notify_from = Column(String(500), name="NOTIFY_FROM")
    papel = Column(String(50), name="PAPEL")
    error = Column(String, nullable=True, name="ERROR")
    enviado = Column(Boolean, name="ENVIADO_API")

    def __repr__(self):
        return "<Comprobante {} clave: {} consecutivo {} estado: {}>".format(
            self.id,
            self.clave,
            self.numero_consecutivo,
            "ENVIADO" if self.enviado else "NO-ENVIADO",
        )

    def auth_data(self) -> dict:
        """ Return a dict with data for authentication
        """
        data = {
            "id": self.emisor_ident_num,
            "entorno": "prueb" if getenv("DEBUG") else "prod",
        }

        return data

    def get_factura(self):
        try:
            return (
                session.query(Factura)
                .filter(Factura.num_factura == self.num_factura and Factura.foliomh == self.referencia_numero)
                .one()
            )
        except MultipleResultsFound:
            raise ValueError("Factura {} esta repetida".format(self.num_factura))

    @staticmethod
    def count():
        """ Return the amount of vouchers to send
        """
        return session.query(Comprobante).filter(Comprobante.enviado == False).count()

    def marshall(self):
        """ Return a dict with all fields
        """
        data = {}
        data["Clave"] = ""
        data["Situacion"] = "1"
        data["NumeroConsecutivo"] = self.numero_consecutivo
        data["FechaEmision"] = self.fecha_emision.isoformat()
        data["CondicionVenta"] = self.condicion_venta
        data["MedioPago"] = self.medio_pago
        data["Emisor"] = {
            "Identificacion": {"Numero": self.emisor_ident_num},
            "Ubicacion": {
                "Provincia": self.emisor_provincia,
                "Canton": self.emisor_canton,
                "Distrito": self.emisor_distrito,
                "OtrasSenas": self.emisor_otras_senas,
            },
            "Telefono": {"CodigoPais": "506"},
        }
        data["Receptor"] = {
            "Nombre": self.receptor_nombre,
            "Identificacion": {
                "Tipo": self.receptor_ident_tipo,
                "Numero": self.receptor_ident_num,
            },
        }
        if self.referencia_numero == None:
            data["DetalleServicio"] = []
            factura = self.get_factura()
            gravado, exento, descuentos, impuestos = 0.0, 0.0, 0.0, 0.0
            for detail in factura.get_details():
                data["DetalleServicio"].append(detail)
                linea = detail["LineaDetalle"]
                if "Impuesto" in linea:
                    gravado = gravado + linea["SubTotal"]
                    impuestos = impuestos + linea["Impuesto"][0]["Monto"]
                else:
                    exento = exento + linea["SubTotal"]
    
                if linea["MontoDescuento"] > 0:
                    descuentos = descuentos + linea["MontoDescuento"]
            # This is hacky. Update the record here.
            # Suma de "SubTotal" de las linea que tengan "Impuesto"
            self.resumen_total_mercancias_gravadas =  round(gravado , 5)
            # Suma de "SubTotal" de las lineas sin "Impuesto"
            self.resumen_total_mercancias_exentas = round(exento, 5) 
            # Igual que `self.resumen_total_mercancias_gravadas`
            self.resumen_total_gravado = round(gravado, 5) 
            # Igual que `self.resumen_total_mercancias_exentas`
            self.resumen_total_exento = round(exento , 5)
            # `self.resumen_total_mercancias_gravadas` + `self.resumen_total_mercancias_exentas`
            self.resumen_total_venta = round(gravado + exento, 5)
            # Suma de 'MontoDescuento' de las lineas de detalle
            self.resumen_total_descuentos = round(descuentos, 5)
            # "TotalVenta" - "TotalDescuentos"
            self.resumen_total_venta_neta = round((gravado + exento) - descuentos , 5)
            # Suma de 'Monto' en 'Impuesto' de todas las lineas de detalle
            self.resumen_total_impuesto = round(impuestos, 5)
            # 'TotalVentaNeta' + 'TotalImpuesto'
            self.resumen_total_comprobante = round((gravado + exento + impuestos) - descuentos , 5)
            # Update the record
            session.commit()
        else:
            pass
        data["ResumenFactura"] = {
            "CodigoMoneda": "CRC",
            "TipoCambio": 1,
            "TotalMercanciasGravadas": "%.5f" % self.resumen_total_mercancias_gravadas,
            "TotalMercanciasExentas": "%.5f" % self.resumen_total_exento,
            "TotalGravado": "%.5f" % self.resumen_total_gravado,
            "TotalExento": "%.5f" % self.resumen_total_exento,
            "TotalVenta": "%.5f" % self.resumen_total_venta,
            "TotalDescuentos": "%.5f" % self.resumen_total_descuentos,
            "TotalVentaNeta": "%.5f" % self.resumen_total_venta_neta,
            "TotalImpuesto": "%.5f" % self.resumen_total_impuesto,
            "TotalComprobante": "%.5f" % self.resumen_total_comprobante,
        }
        if self.referencia_numero == None:
            pass
        else:
            data["InformacionReferencia"] = {
                "Numero" : self.referencia_numero,
            }
        data["Normativa"] = {
            "NumeroResolucion": self.normativa_num_resolucion,
            "FechaResolucion": self.normativa_fecha_resolucion,
        }
        data["Otros"] = self.otros
        data["notify"] = {"to": self.notify_to, "from": self.notify_from}
        data["papel"] = self.papel

        return data
